File: src/clients/memcached-client.js
const Memcached = require('memcached');
const fs = require('fs');

class MemcachedClient {
  
  constructor(numOfInstances = 1, resultFilename = 'results-memcached.txt', totalOperations = 10000) {
    this.config = this.generateConfig(numOfInstances);
    this.numOfInstances = numOfInstances;
    this.resultFilename = resultFilename;
    this.totalOperations = totalOperations;
  }
  
  generateConfig(numOfInstances) {
    let config = [];
  
    for (let i = 0; i < numOfInstances; i++) {
      config.push(`127.0.0.1:1121${i}`);
    }
  
    return config;
  }

  async connect() {
    this.client = new Memcached(this.config, {
      poolSize: this.numOfInstances,
    });
    console.log("Connected to Memcached");
    return this.client;
  }

  async disconnect() {
    await this.cleanup();
    this.client.end();
    console.log("Memcached Client Disconnected");
  }

  benchmarkWrite(dataSize, show=false) {
    return new Promise((resolve) => {
      const data = Buffer.alloc(dataSize, 'a');
      let totalTimeTaken = 0;
      this.keys = [];
      const operations = [];
  
      for (let i = 0; i < this.totalOperations; i++) {
        const key = `key-${dataSize}-${i}`;
        this.keys.push(key);
        
        operations.push(new Promise((resolve, reject) => {
            this.client.set(key, data, 10000, (err) => {
                if (err) reject(err);
                else resolve();
            });
        }));
      }
  
      const start = process.hrtime.bigint();
      Promise.all(operations).then(() => {
        const end = process.hrtime.bigint();
    
        totalTimeTaken += Number(end - start) / 1e6; 
    
        const averageTime = totalTimeTaken / this.totalOperations;
        if(show) {
          console.info(`Memcached WRITE ASYNC for size ${dataSize} bytes: ${averageTime.toFixed(6)} ms per operation`);
          fs.appendFile(this.resultFilename, 'Memcached, WRITE ASYNC, ' + this.numOfInstances + ', ' + dataSize + ', ' + averageTime.toFixed(6) + '\n', (err) => {
            if(err) console.error(err);
          });
        }
    
        resolve();
      });
    })
  }

  benchmarkRead(dataSize, show=false) {
    return new Promise((resolve) => {
      let keys = this.keys
                    .map(val => ({ val, sort: Math.random() }))
                    .sort((a, b) => a.sort - b.sort)
                    .map(obj => obj.val);

      // if(dataSize > 16384) {
      //   keys = keys.slice(0, 1000);
      // }
  
      let operations = [];
  
      operations = keys.map(key => {
        return new Promise((resolve, reject) => {
          this.client.get(key, () => {
              resolve();
          });
        })
      })

      const start = process.hrtime.bigint();
      Promise.all(operations).then(() => {
      // operations[0].then(() => {
        const end = process.hrtime.bigint();
    
        const averageTime = (Number(end - start) / 1e6) / this.keys.length;
        if(show) {
          console.info(`Memcached READ ASYNC for size ${dataSize} bytes: ${averageTime.toFixed(6)} ms per operation`);
          fs.appendFile(this.resultFilename, 'Memcached, READ ASYNC, ' + this.numOfInstances + ', ' + dataSize + ', ' + averageTime.toFixed(6) + '\n', (err) => {
            if(err) console.error(err);
          });
        }
  
        resolve();
      });
    })

  }

  cleanup() {
    return new Promise((resolve) => {
      this.client.flush((err) => {
        if (err) {
          console.error(err);
        }
        // console.log("Memcached Client Cleaned Up");
        resolve();
      })
    });
  }

  shuffle(array) {
    let currentIndex = array.length;
  
    while (currentIndex != 0) {
  
      let randomIndex = Math.floor(Math.random() * currentIndex);
      currentIndex--;
  

      [array[currentIndex], array[randomIndex]] = [
        array[randomIndex], array[currentIndex]];
    }

    return array;
  }
}

module.exports = MemcachedClient;

File: python_clients/memcached_client.py
from pymemcache.client.base import Client as memcachedClient
from pymemcache.client.hash import HashClient as memcachedCluster
from concurrent.futures import ThreadPoolExecutor, as_completed
from multiprocessing import Pool
from multiprocessing import Array
from threading import Thread
import threading
import random
import time

class MemcachedClient:
  def __init__(self, numOfInstances = 1, resultFilename = 'res-memcached-py.txt', totalOperations = 10000, totalProcesses = 1):
    self.numOfInstances = numOfInstances
    self.config = self.generateConfig(self.numOfInstances)
    self.resultFilename = resultFilename
    self.totalOperations = totalOperations
    self.totalProcesses = totalProcesses
    self.keys = []
    self.client = None


  def generateConfig(self, numOfInstances):
    config = []

    config.append('192.168.1.7:11210')
    if numOfInstances > 1: config.append('192.168.1.110:11211')
    if numOfInstances > 2: config.append('192.168.1.9:11212')

    print(config)
    
    return config


  def connect(self):
    if len(self.config) == 1:
      self.client = memcachedClient(self.config[0])
    else:
      self.client = memcachedCluster(self.config)

  def disconnect(self):
    self.client.disconnect_all()


  def benchmarkWrite(self, dataSize, show=False):
    data = 'a' * dataSize

    totalTimeTaken = 0
    for i in range(self.totalOperations):
      key = f'key-{dataSize}-{i}'
      self.keys.append(key)

      start = time.time()
      self.client.set(key, data)
      end = time.time()

      totalTimeTaken += end - start

    averageTime = totalTimeTaken / self.totalOperations
      
    resultFile = open(self.resultFilename, 'a')
    print(f'Memcached WRITE SINGLE {dataSize} {averageTime:.8f}')
    resultFile.write(f'Memcached WRITE SINGLE {dataSize} {averageTime:.8f}\n')
    resultFile.close()

  def benchmarkClusterWrite(self, processNum, totalProcesses):
    self.connect()
    totalTimeTaken = 0
    data = 'a' * (1024 // totalProcesses)

    for i in range(self.totalOperations):
      key = f'key-{i}-{processNum}'
      start = time.time()
      self.client.set(key, data)
      end = time.time()

      totalTimeTaken += end - start

    averageTime = totalTimeTaken / self.totalOperations

    print(f'Memcached WRITE CLUSTER {averageTime:.8f}')
    resultFile = open(self.resultFilename, 'a')
    resultFile.write(f'Memcached WRITE CLUSTER {averageTime:.8f}\n')
    resultFile.close()

    self.disconnect()

  def benchmarkClusterRead(self, processNum, totalProcesses):
    self.connect()
    totalTimeTaken = 0

    for i in range(self.totalOperations):
      key = f'key-{i}-{processNum}'
      start = time.time()
      self.client.get(key)
      end = time.time()

      totalTimeTaken += end - start

    averageTime = totalTimeTaken / self.totalOperations

    print(f'Memcached READ CLUSTER {averageTime:.8f}')
    resultFile = open(self.resultFilename, 'a')
    resultFile.write(f'Memcached READ CLUSTER {averageTime:.8f}\n')
    resultFile.close()

    self.disconnect()


  def benchmarkWriteMultiprocess(self, dataSize, show=False):
    data = 'a' * dataSize

    totalTimeTaken = 0
    for i in range(self.totalOperations):
      key = f'key-{dataSize}-{i}'
      self.keys.append(key)

      start = time.time()
      self.client.set(key, data)
      end = time.time()

      totalTimeTaken += end - start

    averageTime = totalTimeTaken / self.totalOperations
    if show:
      print(f'Memcached WRITE MULTI {dataSize} {averageTime:.8f}')
      
    resultFile = open(self.resultFilename, 'a')
    resultFile.write(f'Memcached WRITE MULTI {dataSize} {averageTime:.8f}\n')
    resultFile.close()

  def benchmarkWriteMultiprocessOld(self, dataSize, offset, show=False, processNum=1):
    self.connect()
    data = 'a' * dataSize

    totalTimeTaken = 0
    for i in range(offset, offset + self.totalOperations//self.totalProcesses):
      key = f'key-{dataSize}-{i}'
      self.keys.append(key)

      start = time.time()
      self.client.set(key, data)
      end = time.time()

      totalTimeTaken += end - start

    averageTime = totalTimeTaken / self.totalOperations
    if show:
      print(f'Memcached WRITE MULTI {dataSize} {averageTime:.8f}')
      
    resultFile = open(self.resultFilename, 'a')
    resultFile.write(f'Memcached WRITE MULTI {dataSize} {averageTime:.8f}\n')
    resultFile.close()
    
    self.disconnect()


  def benchmarkRead(self, dataSize, show=False):
    random.shuffle(self.keys)

    totalTimeTaken = 0
    for i in range(self.totalOperations):
      key = self.keys[i]
      start = time.time()
      self.client.get(key)
      end = time.time()

      totalTimeTaken += end - start

    averageTime = totalTimeTaken / self.totalOperations
    if show:
      print(f'Memcached READ SINGLE {dataSize} {averageTime:.8f}')
      
    resultFile = open(self.resultFilename, 'a')
    resultFile.write(f'Memcached READ SINGLE {dataSize} {averageTime:.8f}\n')
    resultFile.close()


  def benchmarkReadMultiprocess(self, dataSize, show=False, processNum=1):
    self.connect()
    random.shuffle(self.keys)

    totalTimeTaken = 0
    for i in range(self.totalOperations):
      key = self.keys[i]
      start = time.time()
      self.client.get(key)
      end = time.time()

      totalTimeTaken += end - start

    averageTime = totalTimeTaken / self.totalOperations
    if show:
      print(f'Memcached READ MULTI {dataSize} {averageTime:.8f}')
      
    resultFile = open(self.resultFilename, 'a')
    resultFile.write(f'Memcached READ MULTI {dataSize} {averageTime:.8f}\n')
    resultFile.close()

    self.disconnect()


  def cleanup(self):
    self.client.flush_all()



